### ISSUE TITLE
Theoretical underpinning behind using hardmax for probability calculation of word

### ISSUE THREAD

**Issue by ChaiBapchya** [v1n337/movie-dialogue-generation#1]
In the file [Movie-dialogue-generation-tensorflow Python notebook](https://github.com/vineetjohn/movie-dialogue-generation/blob/60d943632c5459aae28c2b5e1073a2801b9fa127/movie-dialogue-generation-tensorflow.ipynb)

While building the computational graph in Tensorflow
```
# Calculate the probability of generating each word
# probs = tf.nn.softmax(logits, name='probs')
probs = tf.contrib.seq2seq.hardmax(logits, name='probs')

```

What's the theoretical underpinning behind using hardmax for calculating probability of generating each word?
What's the advantage of that over other?
How does the gradient calculation work?
